### PR TITLE
Reduce stream

### PR DESCRIPTION
**Pull request by segayuu**
## How to test

```sh
git clone -b BRANCH https://github.com/USER/hexo.git
cd hexo
npm install
npm test
```

## Pull request tasks

- [ ] Add test cases for the changes.
- [ ] Passed the CI test.

### CODE REVIEW

**Comment by SukkaW** [comment]
Whoa, nearly all test failed!

File: lib/plugins/helper/toc.js [hunk 1]
```diff
@@ -3,12 +3,13 @@
 const { tocObj, escapeHTML, encodeURL } = require('hexo-util');
 
 function tocHelper(str, options = {}) {
-  options = Object.assign({
+  options = {
```
Comment (SukkaW) [inline lib/plugins/helper/toc.js:6]: Only when first parameter of `Object.assign` is an empty Object will syntax be faster than `Object.assign`. Otherwise I can't see why should we use spread syntax to shallow merge Object.

File: lib/extend/console.js [hunk 2]
```diff
@@ -5,17 +5,21 @@ const abbrev = require('abbrev');
...
+    return this.store.get(this.alias[name]);
   }
 
   list() {
-    return this.store;
+    const result = {};
+    for (const [key, value] of this.store) {
+      result[key] = value;
+    }
+    return result;
```
Comment (SukkaW) [inline lib/extend/console.js:22]:
```suggestion
    // FIXME: Could use Object.fromEntries after Node.js 10 is dropped.
    const result = {};
    for (const [key, value] of this.store) {
      result[key] = value;
    }
    return result;
```

Could add a `FIXME` message.

File: lib/extend/filter.js [hunk 3]
```diff
@@ -9,12 +9,18 @@ const typeAlias = {
...
-    if (!type) return this.store;
-    return this.store[type] || [];
+    if (!type) {
+      const result = {};
+      for (const [key, value] of this.store) {
+        result[key] = value;
+      }
+      return result;
+    }
+    return this.store.get(type) || [];
```
Comment (SukkaW) [inline lib/extend/filter.js:23]: Same here, a `FIXME` message for future development.

File: lib/extend/generator.js [hunk 4]
```diff
@@ -5,15 +5,19 @@ const Promise = require('bluebird');
...
+    this.store = new Map();
   }
 
   list() {
-    return this.store;
+    const result = {};
+    for (const [key, value] of this.store) {
+      result[key] = value;
+    }
+    return result;
```
Comment (SukkaW) [inline lib/extend/generator.js:16]: Same here, a FIXME message for future development.

File: lib/extend/helper.js [hunk 5]
```diff
@@ -2,22 +2,26 @@
...
+    this.store = new Map();
   }
 
   list() {
-    return this.store;
+    const result = {};
+    for (const [key, value] of this.store) {
+      result[key] = value;
+    }
+    return result;
```
Comment (SukkaW) [inline lib/extend/helper.js:13]: Same here, a FIXME message for future development.

File: lib/extend/migrator.js [hunk 6]
```diff
@@ -4,15 +4,19 @@ const Promise = require('bluebird');
...
+    this.store = new Map();
   }
 
   list() {
-    return this.store;
+    const result = {};
+    for (const [key, value] of this.store) {
+      result[key] = value;
+    }
+    return result;
```
Comment (SukkaW) [inline lib/extend/migrator.js:15]: Same here, a FIXME message for future development.

File: lib/models/types/moment.js [hunk 7]
```diff
@@ -35,7 +31,7 @@ class SchemaTypeMoment extends SchemaType {
   }
 
   match(value, query, data) {
-    return value ? value.valueOf() === query.valueOf() : false;
+    return !!value && value.valueOf() === query.valueOf();
```
Comment (SukkaW) [inline lib/models/types/moment.js:34]:
```suggestion
    return Boolean(value) && value.valueOf() === query.valueOf();
```

Personally I prefer using `Boolean()` approach. But since it is not included in `eslint-config-hexo`, use `!!` will be fine as well.

File: lib/hexo/locals.js [hunk 8]
```diff
@@ -48,10 +48,8 @@ class Locals {
 
   toObject() {
     const result = {};
-    const keys = Object.keys(this.getters);
 
-    for (let i = 0, len = keys.length; i < len; i++) {
-      const key = keys[i];
+    for (const key of this.getters.keys()) {
```
Comment (SukkaW) [inline lib/hexo/locals.js:52]: Should we use `for..of` to replace traditional `for` loop? Last time I checked the `for..of` is nearly 20% slower than traditional `for` loop.

File: lib/plugins/processor/post.js [hunk 9]
```diff
@@ -60,9 +59,8 @@ module.exports = ctx => {
         data.published = false;
       }
 
-      for (let i = 0, len = keys.length; i < len; i++) {
-        const key = keys[i];
-        if (!preservedKeys[key]) data[key] = info[key];
+      for (const key of Object.keys(info)) {
```
Comment (SukkaW) [inline lib/plugins/processor/post.js:62]: Same here. `for..of` is considered to be not so fast as traditional `for` loop.

**Comment by SukkaW** [comment]
Also, the generating performance regression should be aware as well (Total generation time has increased by 23% during the benchmark).

File: lib/hexo/locals.js (hunk 8 above)
Comment (segayuu) [reply lib/hexo/locals.js:52]: The return value of `Map#keys()` is Iterable, not Array. This is different than `Object.keys()`.
It's interesting how fast I get if I wrap it in `Array.from()` and do it in a for loop :)

**Comment by SukkaW** [comment]
The Map is only faster for frequent adding & removals.

For extend APIs (which only have a few registers), it seems better to retain using Object.

File: lib/box/index.js [hunk 10]
```diff
@@ -216,20 +216,17 @@ function escapeBackslash(path) {
...
+  const hasher = createSha1Hash();
+
+  const finishedPromise = new Promise((resolve, reject) => {
+    src.once('error', reject);
+    src.once('end', resolve);
   });
+
+  src.on('data', chunk => { hasher.update(chunk); });
+
+  return finishedPromise.then(() => hasher.digest('hex'));
```
Comment (SukkaW) [inline lib/box/index.js:229]: Would you mind revert the changes here? It seems that the changes caused the performance regression. Let's see if any improvements could be made by reverting changes here

**Comment by SukkaW** [comment]
Also, would you mind splitting into more smaller PR? Too many commits cover too many fields make it very hard to review and locate the performance issue.

**Comment by segayuu** [comment]
Thank you for the PR split. This PR will be reduced according to the title.
